File: go/utils/performance.go
package utils

import (
	"fmt"
	"time"
)

func Wrap(f AdventFunc) AdventFunc {
	wrapped := func(raw string) string {
		return Profile(f, raw)
	}
	return wrapped
}

func Profile(f AdventFunc, raw string) string {
	t1 := time.Now()
	result := f(raw)
	t2 := time.Now()
	dt := float32(t2.UnixMicro()-t1.UnixMicro()) / 1000000
	message := fmt.Sprintf("%v: %v (%.5f seconds)",
		GetFuncName(f),
		result,
		dt,
	)
	fmt.Println(message)
	return result
}

File: python/utils/performance.py
import functools
import time
from typing import Callable, ParamSpec, TypeVar

T = TypeVar("T")
P = ParamSpec("P")


def profile(func: Callable[P, T]) -> Callable[P, T]:
    @functools.wraps(func)
    def wrapped(*args: P.args, **kwargs: P.kwargs):
        t1 = time.perf_counter()
        result = func(*args, **kwargs)
        t2 = time.perf_counter()
        print(f"{func.__name__}: {result} ({t2 - t1:.5f} seconds)")
        return result

    return wrapped
